kissh.py

```diff
@@ -0,0 +1,142 @@
+#!/usr/bin/env python3
+import argparse
+import pwd
+import subprocess
+import sys
+import traceback
+from pathlib import Path
+
+import requests
+
+session = requests.Session()
+
+
+def user_exists(user):
+    try:
+        user_info = pwd.getpwnam(user)
+        return Path(user_info.pw_dir) / ".ssh/authorized_keys"
+    except KeyError:
+        return None
+
+
+def create_user(user):
+    subprocess.run(["adduser", user, "--gecos=''", "--disabled-password"], check=True)
+    subprocess.run(["adduser", user, "sudo"], check=True)
+    # force password set on first ssh login, by deleting and expiring their
+    # passwd
+    subprocess.run(["passwd", "-de", user], check=True)
+
+    return user_exists(user)
+
+
+def get_fingerprint(key):
+    ps = subprocess.run(
+        ["ssh-keygen", "-lf-"],
+        input=key,
+        check=True,
+        capture_output=True,
+        text=True,
+    )
+    # ssh-keygen output format is: "SIZE HASH:VALUE no comment (TYPE)".
+    # eg: "256 SHA256:AsFa6mIf22oiMZSW7yNn3Fip2Ri6DAzjrh+KiZ+axWg no comment (ED25519)"
+    return ps.stdout.split()[1]
+
+
+def get_github_key(user, fingerprint):
+    """Get Github user's key that matches fingerprint."""
+    # Do not use the api to avoid ratelimiting
+    r = session.get(f"https://github.com/{user}.keys")
+    r.raise_for_status()
+
+    # Find the correct key by its fingerprint
+    for key in r.text.splitlines():
+        if get_fingerprint(key) == fingerprint:
+            return key
+
+
+def write_file(path, contents):
+    path.parent.mkdir(exist_ok=True)
+    path.write_text(contents)
+
+
+def manage_user(user, fingerprint):
+    auth_keys = user_exists(user)
+    key = get_github_key(user, fingerprint)
+    if key:
+        if not auth_keys:
+            auth_keys = create_user(user)
+        write_file(auth_keys, key)
+    else:
+        # remove all keys from users authorized_keys
+        if user_info:
```

Comment:
What is `user_info`?

Reply:
huh, left over from previous iteration, it should be auth_keys.

I need to figure out away to unit test this properly. I think I need be able to use a different passwd file, which means I can do the full add/update/remove cycle in the tests.

Reply:
I've added tests for a full walk through of adding a new user, and then changing their key and removing them